File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(num, hand):
    key ={
        1: (0, 0), 2: (0, 1), 3: (0, 2),
        4: (1, 0), 5: (1, 1), 6: (1, 2),
        7: (2, 0), 8: (2, 1), 9: (2, 2),
        '*': (3, 0), 0: (3, 1), '#': (3, 2)
    }
    
    left_pos = key['*']
    right_pos = key['#']
    res = []
    
    for n in num:
        target = key[n]
        
        if n in [1,4,7]:
            res.append('L')
            left_pos = target
        elif n in [3,6,9]:
            res.append('R')
            right_pos = target
        else:
            dist_l = abs(left_pos[0] - target[0]) + abs(left_pos[1] - target[1])
            dist_r = abs(right_pos[0] - target[0]) + abs(right_pos[1] - target[1])
            
            if dist_l < dist_r:
                res.append('L')
                left_pos = target
            elif dist_r < dist_l:
                res.append('R')
                right_pos = target
            else:
                if hand == "left":
                    res.append('L')
                    left_pos = target
                else:
                    res.append('R')
                    right_pos = target
                    
    return "".join(res)

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

pair<int, int> pos[12] = {
    {3,1}, // 0
    {0,0}, {0,1}, {0,2}, // 1,2,3
    {1,0}, {1,1}, {1,2}, // 4,5,6
    {2,0}, {2,1}, {2,2}, // 7,8,9
    {3,0}, // *: 왼손 시작점 10
    {3,2}, // #: 오른손 시작점 11
};

// 맨해튼 거리로 좌표 거리 계산
int get_dist(pair<int,int> a, pair<int,int> b){
    return abs(a.first - b.first) + abs(a.second - b.second);
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    
    pair<int,int> left_pos = pos[10];  // * 위치
    pair<int,int> right_pos = pos[11]; // # 위치
    
    for(int num: numbers){
        if(num == 1 || num == 4 || num == 7){
            answer += 'L';
            left_pos = pos[num];
        }
        else if(num == 3 || num == 6 || num == 9){
            answer += 'R';
            right_pos = pos[num];
        }
        else{
            int left_dist = get_dist(left_pos, pos[num]);
            int right_dist = get_dist(right_pos, pos[num]);
            
            if (left_dist < right_dist){
                answer += 'L';
                left_pos = pos[num];
            }
            else if(right_dist < left_dist){
                answer += 'R';
                right_pos = pos[num];
            }
            else{
                if (hand == "left"){
                    answer += 'L';
                    left_pos = pos[num];
                }
                else{
                    answer += 'R';
                    right_pos = pos[num];
                }
            }
        }
    }
    return answer;
}